src/poly/commitment.rs

```diff
@@ -21,6 +23,95 @@ pub struct OpeningProof<C: CurveAffine> {
     z2: C::Scalar,
 }
 
+/// A multiscalar multiplication in the polynomial commitment scheme
+#[derive(Debug)]
+pub struct MSM<C: CurveAffine> {
+    /// TODO: documentation
+    pub g_scalars: Option<Vec<C::Scalar>>,
+
+    /// TODO: documentation
+    pub h_scalar: Option<C::Scalar>,
+
+    /// TODO: documentation
+    pub other_scalars: Vec<C::Scalar>,
+
+    /// TODO: documentation
+    pub other_bases: Vec<C>,
+}
+
+impl<'a, C: CurveAffine> MSM<C> {
+    /// Empty MSM
+    pub fn default(params: &Params<C>) -> Self {
+        let g_scalars = Some(vec![C::Scalar::one(); params.n as usize]);
+        let h_scalar = Some(C::Scalar::one());
+        let other_scalars: Vec<C::Scalar> = Vec::with_capacity(params.k as usize * 2 + 3);
+        let other_bases: Vec<C> = Vec::with_capacity(params.k as usize * 2 + 3);
+
+        MSM {
+            g_scalars,
+            h_scalar,
+            other_scalars,
+            other_bases,
+        }
+    }
+
+    /// Add arbitrary term (the scalar and the point)
+    pub fn add_term(&mut self, scalar: C::Scalar, point: C) {
+        &self.other_scalars.push(scalar);
+        &self.other_bases.push(point);
+    }
+
+    /// Add a vector of scalars to `g_scalars`
+    pub fn add_to_g(&mut self, scalars: Vec<C::Scalar>) {
+        for (g_scalar, scalar) in self
```

Comment:
This code should check to see if `self.g_scalars` is `None`, and if so it should set it to `Some(scalars.to_vec())`. Otherwise in case it is `Some` then you can perform the loop to perform the addition as you've written it.